Add UI for realm-level defaults of user display settings.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Commit summary -
- First two commits are to hide certain settings in realm-default section.
- Next four commits are prep commits to extract `settings_display.js` such that we can avoid duplication of code.
- Next commit is to add `realm_user_settings_defaults` module similar to `user_settings` module.
- Last commit is to add UI and javascript logic for realm-level defaults of display settings.
 <!-- How have you tested? -->

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![Screenshot from 2021-09-12 23-12-59](https://user-images.githubusercontent.com/35494118/132997572-5ba077a7-7fc4-4063-9f0e-e23904349f3b.png)



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented:
A couple of things -
- I am not sure what icon to use for the `User settings defaults` section. For now I have used the cog icon. I thought of using the [user-cog icon](https://fontawesome.com/v5.15/icons/user-cog?style=solid) but it is not available for the version we use of Font awesome.
- Any suggestions regarding the name of files, especially `settings_defaults.js`.

timabbott commented:
This generally looks pretty clean!  I merged the first few commits as the series ending with 0c11ced0a0cac8541f4a733fee01416241496d36.

* I think cog looks good as a starting point; we can do a quick `#design` thread to finalize that decision once the UI is merged and it's easy to play with.
* `settings_defaults.js` might be better `settings_realm_user_settings_defaults`; it's a bit of a mouthful and kinda repetitive but we're only going to reference it a few times, and it fits the convention we have with a `foo` data module and a `settings_foo` module for settings UI to change that state.
* Posted https://github.com/sahil839/zulip/commit/2999a003a17a2a3b0037cfc7067863f5b9c2b095#r56387195 (which sadly doesn't appear here) for the `initialize` function naming.
* Do we need the `user_display_settings` module?  It seems like it might be simpler to just have the `settings_sections` code register `() => {settings_display.set_up($("#user-display-settings"), user_settings)}` and avoid the extra layer of indirection.
* For `settings: Make separate object for display settings labels.`, should we put that object in `settings_config.ts` above the notifications settings version of that?  Seems a bit cleaner.
* I think we probably need to add or extend the tip at the top of the page to explain what these are: "Configure the default personal preference settings for new users joining your organization" before we can merge the final UI commit.

sahil839 commented:
Updated the PR with following changes-
- Renamed `settings_defaults.js` to `settings_realm_user_settings_defaults.js`.
- Deleted `user_display_settings.js` module and instead modified the code in `settings_section.js`. We would need `settings_realm_user_settings_defaults.js` since some settings like `enter_sends` and `presence_enabled` are neither in display settings nor notification settings UI.
- Moved `display_settings_label` object to `settings_config`.
- Renamed the initialization function to `initialize` in `realm_user_settings_defaults.ts`. I first thought it to be similar to `user_settings` but we actually do not access the `realm_user_settings_defaults` object in `ui_init.js` so ther would be no conflicts.

sahil839 commented:
Regarding the tip in UI, one option is that we can use the yellow box that we use in other sections or we can simply write text but I think if we want to write text we should add a top heading like `Default user settings` and then add the tip text and then the existing `Display settings` heading and so on. Attaching some screenshots below. Would love to have some other suggestions other than these if any.

![Screenshot from 2021-09-14 00-31-32](https://user-images.githubusercontent.com/35494118/133144489-b02a3afc-fb8c-4d43-9596-745539635526.png)

![Screenshot from 2021-09-14 00-32-58](https://user-images.githubusercontent.com/35494118/133144501-5c830432-56ee-4de2-9be5-505729c9cd0f.png)

![Screenshot from 2021-09-14 00-28-14](https://user-images.githubusercontent.com/35494118/133144505-b339c901-96eb-4aa1-a466-d95d5d9b764c.png)

sahil839 commented:
Also apart from the above tip, do we need tip like "Only admins can edit these settings" like we have for other pages.
If yes, should this tip be shown to admins also who are actually allowed to edit (asking this because this is not consistent for all sections).

timabbott commented:
@alya and I agree the yellow tip is the way to go among those screenshots.  I think it's OK to have two of these boxes, e.g. "bot settings" has that in some situations.

I think this is ready to merge with the following changes:
* Change the user-facing label for the panel to "Default user settings" -- that feels a bit cleaner.
* Add the tip described above (can be in a new final commit to avoid re-review of the rest)

As short-term follow-ups, I think the following follow-ups are important:
* We ideally would use the save/discard UI for the "organization settings" version of these pages, to be consistent with how all of our other administrator-level settings work.  I don't recall how that works in the HTML and thus how much refactoring is required for us to have a save/discard widget share a template with a basic widget.
* As you mention, adding the notification settings and "untyped settings" to this page.
* Moving the language/time settings to use this system.
* Add /help/ documentation for this, which @alya can probably drive.  This is probably easier to write after the above are done.

And here's a few potentially longer-term ideas:
* It would probably be nice if we can display which value is the default for our settings, since I can readily imagine an administrator wanting to know what they've changed from Zulip's global defaults.  I think this is probably only worth working on now if we see a straightforward way to do it, since it's not super high value and might be a lot of work.

sahil839 commented:
@timabbott updated the user-facing label in left-panel and added tip.
![Screenshot from 2021-09-15 19-55-38](https://user-images.githubusercontent.com/35494118/133453487-ece0f147-fb9e-439e-812b-723d14c4a6de.png)

sahil839 commented:
Also, we need the save-discard widget only for realm default settings and not for user-settings right?
Though, we could not use a lot of code in `settings_display.js` then and there would be some `if-else` conditions needed as we use common template too.

timabbott commented:
Merged after adding a `.` at the end of the tip string and adding a bonus commit to fix the fact that the "bot settings" string was our one existing tip string that didn't end with a period.  Thanks @sahil839!

For the save/discard widget: the difference in HTML structure is basically this:

```
diff --git a/static/templates/settings/display_settings.hbs b/static/templates/settings/display_settings.hbs
index 1a175f7f0d..de46b25c87 100644
--- a/static/templates/settings/display_settings.hbs
+++ b/static/templates/settings/display_settings.hbs
@@ -19,8 +19,10 @@
     {{/unless}}
 
     <div class="display-settings">
-        <h3 class="inline-block">{{t "Display settings" }}</h3>
-        <div class="alert-notification display-settings-status"></div>
+        <div class="subsection-header">
+            <h3>{{t "Display settings" }}</h3>
+            {{> settings_save_discard_widget section_name="display-settings" }}
+        </div>
 
         <div class="input-group">
             <label for="default_view" class="dropdown-title">{{t "Default view" }}
```

I think theoretically it should be possible to refactor things so that we use the `section-header` block HTML structure for personal user settings, and effectively the ``settings_save_discard_widget template could then be extended to take a parameter for whether to do save/discard or just the "saving animation".

And then we'd need a similar refactor on the JavaScript side to add a new abstraction where we could replace code like:
```
    container.find(".setting_demote_inactive_streams").on("change", function () {                        
        const data = {demote_inactive_streams: this.value};                                              
        change_display_setting(data, container, patch_url, ".display-settings-status");                  
    });                                                                                                  
```

with an interface that looked more like this:

```
    register_setting_save_function(container.find(".setting_demote_inactive_streams"), function () {                        
        const data = {demote_inactive_streams: this.value};                                              
        change_display_setting(data, container, patch_url, ".display-settings-status");                  
    });                                                                                                  
```

And then that `register` function could setup a save/discard element or an auto-save element depending on the `for_realm_settings` parameter passed into it.  

I haven't thought too deeply about how this should work, but the path to it probably starts with generalizing the `settings_org.js` infrastructure to support both styles of on-change indicators, and then porting `settings_display.js` to use that framework (since the `settings_org.js` stuff already has helpful systems for details like finding the status element and type from the HTML structure rather than needing to pass hardcode it in the JS files).

sahil839 commented:
Ok, will look into this. 
And I was thinking, we can add the other settings first to the UI since I have that work almost ready and then we can do the save discard widget change.

timabbott commented:
Yeah, I think that's the right plan as well.
